corehq/apps/cloudcare/static/cloudcare/js/form_entry/form_ui.js

```diff
@@ -312,22 +315,52 @@ hqDefine("cloudcare/js/form_entry/form_ui", function () {
             usedWidth = 0;
         }
 
-        for (let child of json.children) {
-            if (child.type === constants.QUESTION_TYPE) {
-                const questionTileWidth = Question.calculateColumnWidthForPerRowStyle(child.style);
+        function pushChildAndResetGroup(child) {
+            newChildren.push(child);
+            resetCurrentGroup();
+        }
+
+        function groupElementsAndResetGroup(child) {
+            const newGroup = Container.groupElements(child, elementType);
+            newChildren.push(newGroup);
+            resetCurrentGroup();
+        }
+
+        function addToGroupForAllowedWidth(child) {
+            const questionTileWidth = Question.calculateColumnWidthForPerRowStyle(child.style);
+            usedWidth += questionTileWidth;
+            if (usedWidth > constants.GRID_COLUMNS) {
+                resetCurrentGroup();
                 usedWidth += questionTileWidth;
-                if (usedWidth > constants.GRID_COLUMNS) {
-                    resetCurrentGroup();
-                    usedWidth += questionTileWidth;
-                }
-                addToCurrentGroup(child);
+            }
+            addToCurrentGroup(child);
+        }
+
+        function processChildForGroupingQuestions(child, usedWidth, currentGroup, newChildren) {
+            if (child.type === constants.QUESTION_TYPE) {
+                addToGroupForAllowedWidth(child);
             } else if (child.type === constants.GROUP_TYPE || child.type === constants.REPEAT_TYPE) {
-                const newGroup = Container.groupQuestions(child);
-                newChildren.push(newGroup);
-                resetCurrentGroup();
+                groupElementsAndResetGroup(child);
             } else {
-                newChildren.push(child);
-                resetCurrentGroup();
+                pushChildAndResetGroup(child);
+            }
+        }
+
+        function processChildForGroupingGroups(child, usedWidth, currentGroup, newChildren) {
+            if (child.type === constants.GROUP_TYPE) {
+                addToGroupForAllowedWidth(child);
+            } else if (child.type === constants.REPEAT_TYPE) {
+                groupElementsAndResetGroup(child);
+            } else {
+                pushChildAndResetGroup(child);
+            }
+        }
+
+        for (let child of json.children) {
+            if (elementType === constants.QUESTION_TYPE) {
+                processChildForGroupingQuestions(child, usedWidth, currentGroup, newChildren);
```

Comment:
Is there a way we can reduce the number of sub-functions needed? I had a hard time following in particular the small functions like `groupElementsAndResetGroup` and `pushChildAndResetGroup`.

 I don't think we need two separate processes. I'm imagining this can done as a DFS with the base cases handling the different `elementTypes`/ if there isn't a `n-per-row` style and recursively calling if needed. I haven't fully thought this out so open to if you don't think it's practical.